infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/DirectKubernetesAPIAccessHelper.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.Map;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.Response;
+import okhttp3.Call;
+import okhttp3.Headers;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.ResponseBody;
+import okio.BufferedSink;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.commons.annotation.Nullable;
+
+public class DirectKubernetesAPIAccessHelper {
+  private DirectKubernetesAPIAccessHelper() {}
+
+  /**
+   * This method just performs an HTTP request of given {@code httpMethod} on an URL composed of the
+   * {@code masterUrl} and {@code relativeUri} using the provided {@code httpClient}, optionally
+   * sending the provided {@code body}.
+   *
+   * @param masterUrl the base of the final URL
+   * @param httpClient the HTTP client to perform the request with
+   * @param httpMethod the HTTP method of the request
+   * @param relativeUri the relative URI that should be appended ot the {@code masterUrl}
+   * @param body the body to send with the request, if any
+   * @return the HTTP response received
+   * @throws InfrastructureException on failure to validate or perform the request
+   */
+  public static Response call(
+      String masterUrl,
+      OkHttpClient httpClient,
+      String httpMethod,
+      URI relativeUri,
+      @Nullable HttpHeaders headers,
+      @Nullable InputStream body)
+      throws InfrastructureException {
+    if (relativeUri.isAbsolute() || relativeUri.isOpaque()) {
+      throw new InfrastructureException(
+          "The direct infrastructure URL must be relative and not opaque.");
+    }
+
+    try {
+      URI fullUrl = new URI(masterUrl).resolve(relativeUri);
+      javax.ws.rs.core.MediaType mediaTypeHeader = headers.getMediaType();
+      String mediaType =
+          mediaTypeHeader == null ? "application/json;charset=utf-8" : mediaTypeHeader.toString();
```

Comment:
I would rather use `javax.ws.rs.core.MediaType` to construct the media type, then hardcoded string.

Reply:
Since I need to convert that to `okhttp` header, I don't see much reason.

Reply:
I don't like the header in hardcoded string. What about something like this?
```
String mediaType =
          headers.getMediaType() == null ?
              APPLICATION_JSON_TYPE.withCharset(StandardCharsets.UTF_8.name()).toString() :
              headers.getMediaType().toString();
```

Reply:
I would hazard to say that the API of the MediaType and StandardCharsets types is more likely to change than the mime format. I'm all for indirection but in this case I don't think it is warranted. The hardcoded string is IMHO both more stable and readable than the API-generated alternative.